core/arch/arm/plat-sam/freq.c

```diff
@@ -12,8 +12,102 @@
 #include <kernel/panic.h>
 #include <libfdt.h>
 
+#define OPP_RATES_MAX 8
+
+/* the clock and its rates for CPU OPP */
+struct clk_rates {
+	struct clk *clk;
+	size_t rate_num; /* the number of valid clock rates in @rates */
+	unsigned long rates[OPP_RATES_MAX];
+};
+
+static struct clk_rates *opp_rates;
 static unsigned long freq;
 
+static TEE_Result get_rates_array(struct clk *clk, size_t start_index,
+				  unsigned long *rates, size_t *nb_elts)
+{
+	if (!opp_rates)
+		panic("Invalid CPU OPP Rates Array");
+
+	if (opp_rates->clk != clk)
+		panic("Invalid clock");
+
+	if (!rates) {
+		*nb_elts = opp_rates->rate_num;
+
+		return TEE_SUCCESS;
+	}
+
+	if (start_index > opp_rates->rate_num ||
+	    start_index + *nb_elts > opp_rates->rate_num) {
+		EMSG("Bad parameter(s): start_index %d, nb_elts %d",
+		     start_index, *nb_elts);
+
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	memcpy(rates, &opp_rates->rates[start_index],
+	       *nb_elts * sizeof(*rates));
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result dt_get_opp_hz(const void *fdt, int node, unsigned long *value)
+{
+	const char *property = "opp-hz";
+	const fdt64_t *p = NULL;
+	int len = 0;
+
+	p = fdt_getprop(fdt, node, property, &len);
+	if (!p)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (len != sizeof(*p))
+		return TEE_ERROR_BAD_FORMAT;
+
+	*value = fdt64_ld(p);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result opp_rates_setup(const void *fdt, int node, struct clk *clk)
+{
+	const char *compatible = "operating-points-v2";
+	const struct fdt_property *prop = NULL;
+	unsigned long rate = 0;
+	int opp_table = 0;
+	int offset = 0;
+	int len = 0;
+
+	prop = fdt_get_property(fdt, node, compatible, &len);
+	if (!prop)
+		return TEE_SUCCESS;
+
+	opp_rates = calloc(1, sizeof(*opp_rates));
+	if (!opp_rates) {
+		EMSG("Fail to alloc opp_rates");
+
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+	memset(opp_rates, 0, sizeof(*opp_rates));
+
+	offset = (int)prop - (int)fdt;
+	if (fdt_node_check_compatible(fdt, offset, compatible))
+		panic("Invalid opp-table");
+
+	fdt_for_each_subnode(opp_table, fdt, offset) {
+		if (dt_get_opp_hz(fdt, opp_table, &rate) != TEE_SUCCESS)
+			panic("Get opp-hz failed");
+
+		opp_rates->rates[opp_rates->rate_num++] = rate;
+	}
+	opp_rates->clk = clk;
+	((struct clk_ops *)clk->ops)->get_rates_array = get_rates_array;
```

Comment:
This doesn't look right. Why isn't `clk->ops` assigned correctly from the start?

Reply:
Yes, it's better that `clk->ops` be assigned in `static const struct clk_ops sam9x60_div_pll_ops_chg` in [clk-sam9x60-pll.c](https://github.com/OP-TEE/optee_os/blob/master/core/drivers/clk/sam/clk-sam9x60-pll.c#L389) in this case. But the CPU clock in use is defined in [the device tree](https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/dts/sama7g5.dtsi#L33), it might be another clock so the assignment needs to be changed accordingly.

Due to the frequency of CPU is handled in this file I added the code of getting rates for OPP in this file too instead of into a file for clock. Do you have any suggestions? Thank you.

PS: As `ops` is defined as [`const struct clk_ops *`](https://github.com/OP-TEE/optee_os/blob/master/core/include/drivers/clk.h#L39), `(struct clk_ops *)` is added here for converting the type to avoid the compile error.

Reply:
@etienne-lms can you help us, please?

Reply:
Then maybe it should be a dedicated clock with your pll as its parent?
What do you think @etienne-lms

Reply:
I agree with @GseoC, it would be better this opp-v2 node is a clock provider that registers a clock with parent clock being `clk` and with a `.set_rate` handler that ensures the requested clock rate matches the supported ones. In that scope, your SCMI server should expose this cpu_opp clock.

But... the clk framework will not allow that since it holds a mutex preventing a clock handle to call clk API functions as `clk_set_rate()`.
If this piece of code can ensure this added "cpu_opp" clock has no parents, its `.set_rate` handler could call the `parent->ops->set_rate` handler straight once it has checked the request clock is valid.
Something like:
```c
static TEE_Result cpu_opp_clk_set_rate(struct clk *clk, unsigned long rate, unsigned long parent_rate __unused)
{
	// check @rate is valid

	assert(clk->parent && !clk->parent->parent);
	return clk->parent->ops->set_rate(clk->parent, rate, 0);
}

static struct clk_ops cpu_opp_clk_ops = {
	.set_rate = cpu_opp_clk_set_rate,
	.get_rates_array = get_rates_array,
};

(...)

static TEE_Result get_freq_from_dt(void)
{
	struct clk *cpu_opp_clk = NULL;
	(...)

	cpu_opp_clk = clk_alloc("cpu-opp", &cpu_opp_clk_ops, &clk, 1);
	// check error case

	if (opp_rates_setup(fdt, node, clk)) {
		// erro case
	}

	res = clk_register(cpu_opp_clk);
	(...)

	/* CPU clock is likely always enabled so set its refcount */
	if (clk_enable(cpu_opp_clk))
		panic();
}

Reply:
@etienne-lms Thank you for your suggestions.
Now a new clock for CPU OPP is created if OPP is available and the clock is registered as `AT91_SCMI_CLK_CPU_OPP` in SCMI clock list. Most of the code is moved to a new file `core/drivers/clk/sam/at91_cpu_opp.c`.